ecc_enclave/enclave/shim.h

```diff
@@ -6,27 +6,130 @@
 
 #pragma once
 
+#include <stdbool.h>
 #include <map>
-#include <set>
 #include <string>
 #include <vector>
 
-typedef std::map<std::string, std::string> write_set_t;
-typedef std::set<std::string> read_set_t;
-typedef std::map<void*, std::pair<read_set_t*, write_set_t*>> context_t;
+typedef void* fpc_ctx_t;
 
-// shim put/get
-void get_state(const char* key, uint8_t* val, uint32_t max_val_len, uint32_t* val_len, void* ctx);
-void put_state(const char* key, uint8_t* val, uint32_t val_len, void* ctx);
+// TODO: Some Meta-questions
+// - should we prefix functions with fpc_ or alike to make them more distinct?
+//   Alternatively, we might go C++ and define a shim object like the go shim
+//   (but then maybe we do not necessarily want to force folks to go to C++ and
+//    still enable C?)
+
+// Function which FPC chaincode has to implement
+// ==================================================
+int invoke(const char* args,
+    uint8_t* response,
+    uint32_t max_response_len,
+    uint32_t* actual_response_len,
+    fpc_ctx_t ctx);
+
+/*
+  TODO: compatibility with fabric programming model, we might want to add also 
+
+int init(const char* args,
```

Comment:
yes that would be nice; however, if we are going to support init we may need to reconsider the enclave registration protocol; as we need an registered enclave to perform init; moreover, with the new lifecycle we can also say init is not supported as one can define this in a chaincode definition

Reply:
why is this not supported?  In new lifecycle you can specify init to be optional or not, but i don't see why this would be ruled out?  Obviously you have to run it after enclave registration (and key-management). It's true with current chaincode type it's a bit messy but i think even there it would work: we could (and should) add the setup calls also to the peer instantiate wrapper and then we can call init also from the wrapper. for the go cc of course the init would look then like an invoke but that would be all under the cover?